src/base/associative_cache.hh

```diff
@@ -0,0 +1,339 @@
+     * This is required to update the replacement information data.
+     * @param addr key to the entry
+     * @return The entry if it exists
+     */
+    virtual Entry* accessEntry(const Addr addr)
+    {
+        auto entry = findEntry(addr);
+
+        if (entry) {
+            replPolicy->touch(entry->replacementData);
+        }
+
+        return entry;
+    }
+
+    /**
+     * Update the replacement information for an entry
+     * @param Entry to access and upate
+     */
+    virtual void accessEntry(Entry *entry)
+    {
+        replPolicy->touch(entry->replacementData);
+    }
+
+    /**
+     * Find an entry within the set
+     * @param addr key element
+     * @return returns a pointer to the wanted entry or nullptr if it does not
+     *  exist.
+     */
+    virtual Entry* findEntry(const Addr addr) const
+    {
+        auto tag   = getTag(addr);
+
+        auto entries = indexingPolicy->getPossibleEntries(addr);
+
+        for (auto it = entries.begin(); it != entries.end(); it++) {
+            Entry *entry = static_cast<Entry*>(*it);
+            if (entry->matchTag(tag)) {
+                return entry;
+            }
+        }
+
+        return nullptr;
+    }
+
+    /**
+     * Find a victim to be replaced
+     * @param addr key to select the possible victim
+     * @result entry to be victimized
+     */
+    virtual Entry* findVictim(const Addr addr)
+    {
+        auto candidates = indexingPolicy->getPossibleEntries(addr);
+
+        auto victim = static_cast<Entry*>(replPolicy->getVictim(candidates));
+
+        invalidate(victim);
+
+        return victim;
+    }
+
+    /**
+     * Invalidate an entry and its respective replacement data.
+     *
+     * @param entry Entry to be invalidated.
+     */
+    virtual void invalidate(Entry *entry)
+    {
+        entry->invalidate();
+        replPolicy->invalidate(entry->replacementData);
+    }
+
+    /**
+     * Indicate that an entry has just been inserted
+     * @param addr key of the container
+     * @param entry pointer to the container entry to be inserted
+     */
+    virtual void insertEntry(Addr addr, Entry *entry)
+    {
+        entry->insert(indexingPolicy->extractTag(addr));
+        replPolicy->reset(entry->replacementData);
+    }
+
+    /**
+     * Find the set of entries that could be replaced given
+     * that we want to add a new entry with the provided key
+     * @param addr key to select the set of entries
+     * @result vector of candidates matching with the provided key
+     */
+    std::vector<Entry *>
+    getPossibleEntries(const Addr addr) const
+    {
+        std::vector<ReplaceableEntry *> selected_entries =
+            indexingPolicy->getPossibleEntries(addr);
+        std::vector<Entry *> entries(selected_entries.size(), nullptr);
+
+        unsigned int idx = 0;
+        for (auto &entry : selected_entries) {
+            entries[idx++] = static_cast<Entry *>(entry);
+        }
+
+        return entries;
+    }
```

Comment:
Hi, I would love this method to not be "confined" to the use of replaceable entries only if possible

The Arm TLB supports partial translation entries, meaning a cached translation could store an intermediate step in the translation walk. This leads to a TLB that stores more than one valid entry.. Needless to say the lookup logic prioritizes complete translations over partial ones.. For this reason we would make use of the getPossibleEntries over the findEntry, which returns a single entry only and returns the first entry that matches the TAG....